pkg/schemaconv/openapi.go

```diff
@@ -0,0 +1,255 @@
+		// compute the type inline
+		c2 := c.push("inlined in "+c.currentName, &tr.Inlined)
+		c2.preserveUnknownFields = c.preserveUnknownFields
+		c2.visitSpec(specSchema)
+		c.pop(c2)
+
+		if tr == (schema.TypeRef{}) {
+			// emit warning?
+			tr = schema.TypeRef{
+				NamedType: &untypedName,
+			}
+		} else if tr.Inlined == deducedDef.Atom {
+			// Deduplicate deducedDef (mostly for testing)
+			tr = schema.TypeRef{
+				NamedType: &deducedName,
+			}
+		}
+	}
+
+	return tr
+}
+
+func (c *convert) parseObject(m *spec.Schema) *schema.Map {
+	var fields []schema.StructField
+	for name, member := range m.Properties {
+		fields = append(fields, schema.StructField{
+			Name:    name,
+			Type:    c.makeOpenAPIRef(&member),
+			Default: member.Default,
+		})
+	}
+
+	// AdditionalProperties informs the schema of any "unknown" keys
+	// Unknown keys are enforced by the ElementType field.
+	elementType := func() schema.TypeRef {
+		if m.AdditionalProperties == nil {
+			// According to openAPI spec, an object without properties and without
+			// additionalProperties is assumed to be a free-form object.
+			if c.preserveUnknownFields || len(m.Properties) == 0 {
+				return schema.TypeRef{
+					NamedType: &deducedName,
+				}
+			} else {
+				// If properties are specified, do not implicitly allow unknown
+				// fields
+				return schema.TypeRef{}
+			}
+		} else if m.AdditionalProperties.Schema != nil {
+			// Unknown fields use the referred schema
+			return c.makeOpenAPIRef(m.AdditionalProperties.Schema)
+
+		} else if m.AdditionalProperties.Allows {
+			// A boolean instead of a schema was provided. Deduce the
+			// type from the value provided at runtime.
+			//
+			// This differs from old proto.Models implementation which
+			// left this as incomplete
+			return schema.TypeRef{
+				NamedType: &deducedName,
+			}
+		} else {
+			// Additional Properties are explicitly disallowed by the user.
+			// Ensure element type is empty.
+			return schema.TypeRef{}
+		}
+	}()
+
+	relationship, err := getMapElementRelationship(m.Extensions)
+	if err != nil {
+		c.reportError(err.Error())
+	}
+
+	return &schema.Map{
+		Fields:              fields,
+		ElementRelationship: relationship,
+		ElementType:         elementType,
+	}
+}
+
+func (c *convert) parseList(m *spec.Schema) *schema.List {
+	relationship, mapKeys, err := getListElementRelationship(m.Extensions)
+	if err != nil {
+		c.reportError(err.Error())
+	}
+	elementType := func() schema.TypeRef {
+		if m.Items != nil {
+			if m.Items.Schema == nil || m.Items.Len() != 1 {
+				c.reportError("structural schema arrays must have exactly one member subtype")
+				return schema.TypeRef{
+					NamedType: &deducedName,
+				}
+			}
+
+			subSchema := m.Items.Schema
+			if subSchema == nil {
+				subSchema = &m.Items.Schemas[0]
+			}
+			return c.makeOpenAPIRef(subSchema)
+
+		}
+
+		// A list with no items specified is treated as arbitrary.
+		return schema.TypeRef{
+			NamedType: &untypedName,
+		}
```

Comment:
Shouldn't we only allow that if `type` wasn't specified? I feel like if `type: array`, `items` is required

Reply:
Yeah I can make that log an error, but still this function should always return a value. Now that I'm looking at it it seems I have different treatments of each error:

https://github.com/kubernetes/kube-openapi/blob/f065637e0d18a9f594ea3a68afbfc174778ff295/pkg/schemaconv/openapi.go#L230-L232

This case returns a deduced value. I should reorganize this function to avoid mistakes like that.

What should the `items` be inferred to be in case of an error? I vote untyped makes the most sense?

Reply:
not sure if `type` is unspecified, should array items be `untyped` or `deduced`?

when I changed it the schema for JSONSChemaPropsOrBool (and a few related others) changed:

```
    "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaPropsOrArray": {
      "description": "JSONSchemaPropsOrArray represents a value that can either be a JSONSchemaProps or an array of JSONSchemaProps. Mainly here for serialization purposes."
    },
```